docs/how-to-guides/effective-gno.md

```diff
@@ -0,0 +1,652 @@
+approach allows you to leverage the composability of Gno to build more powerful
+and flexible applications.
+
+### Design your realm as a public API
+
+In Go, all your packages, including your dependencies, are typically treated as
+part of your safe zone, similar to a secure perimeter. The boundary is drawn
+between your program and the rest of the world, which means you secure the API
+itself, potentially with authentication middlewares.
+
+However, in Gno, your package is the public API. It's exposed to the outside
+world and can be accessed by other realms. Therefore, it's crucial to design
+your realm with the same level of care and security considerations as you would
+a public API.
+
+One approach is to simulate a secure perimeter within your realm by having
+private functions for the logic and then writing your API layer by adding some
+front-facing API with authentication. This way, you can control access to your
+realm's functionality and ensure that only authorized callers can execute
+certain operations.
+
+```go
+import "std"
+
+func PublicMethod(nb int) {
+    caller := std.GetOrigCaller()
+    privateMethod(caller, nb)
+}
+
+func privateMethod(caller std.Address, nb int) { /* ... */ }
+```
+
+In this example, `PublicMethod` is a public function that can be called by other
+realms. It retrieves the caller's address using `std.GetOrigCaller()`, and then
+passes it to `privateMethod`, which is a private function that performs the
+actual logic. This way, `privateMethod` can only be called from within the
+realm, and it can use the caller's address for authentication or authorization
+checks.
+
+### Contract-level access control
+
+In Gno, it's a good practice to design your contract as an application with its
+own access control. This means that different endpoints of your contract should
+be accessible to different types of users, such as the public, admins, or
+moderators.
+
+The goal is usually to store the admin address or a list of addresses
+(`std.Address`) in a variable, and then create helper functions to update the
+owners. These helper functions should check if the caller of a function is
+whitelisted or not.
+
+Let's deep dive into the different access control mechanisms we can use:
+
+One approach is to look at the EOA (Externally Owned Account), which is the
+original caller. For this, you should call `std.GetOrigCaller()`, which returns
+the address of the wallet used to make the transaction.
+
+Internally, this call will look at the frame stack, which is basically the stack
+of callers including all the functions, anonymous functions, other realms, and
+take the initial caller. This allows you to identify the original caller and
+implement access control based on their address.
+
+Here's an example:
+
+```go
+import "std"
+
+var admin std.Address = "g1xxxxx"
+
+func AdminOnlyFunction() {
+    caller := std.GetOrigCaller()
+    if caller != admin {
```

Comment:
!!! bad

allows for "gno.land/r/demo/evil" to call "gno.land/r/demo/mycontract".AdminOnlyFunction, and succeed if the origcaller is the admin.

We should recommend PrevRealm().Addr().

Reply:
I can think of very few cases where `GetOrigCaller` actually makes sense to be used. But we're having this discussion partly in #1048. I think we should gliss over GetOrigCaller as we're finalizing details, I think it's currently more sensible to encourage only PrevRealm...

Reply:
also GetOrigCaller auth prevents abstracted accounts, daos and every other "user-like realms" from using functionalities guarded by it

Reply:
When discussing this, we should just see how Ethereum does it. 

You've 2 variables in the execution context of Ethereum:
- msg.sender - the immediate, previous caller:
EOA calls contract A, which calls contract B - In A, msg.sender is EOA, in B, msg.sender is A.

- tx.origin - the real, initial signer of the transaction - in the above case, tx.origin would always be the EOA.

I think while we do differ from Ethereum, we should not reinvent the wheel (too much), so I propose we go with this logic, which I think aligns with most people's views already.

Let's just define this, utilize good naming conventions, be consistent, and add it to the docs.

Reply:
isn't tx.origin here because of backward compatiblity?
I think we should think about this and not do cargo cult mentality actually
can someone explain cases where we do need tx.origin?

Reply:
I recommend using `PrevRealm` in all of Gno, but let's also keep a part about how to use `GetOrigCaller` properly.

Also, we could have `gno lint` give a warning if someone uses `GetOrigCaller`. This warning could stop if you write a comment like `// nolint: I know what I’m doing.`

Reply:
Fixed in 16984a26.